File: Dsa-day180.java
class Solution {
    public double separateSquares(int[][] squares) {
        long total = 0;
        double maxY = 0;
        for (int[] sq : squares) {
            long l = sq[2];
            total += l * l;
            maxY = Math.max(maxY, sq[1] + l);
        }
        double half = total / 2.0;

        double lo = 0, hi = maxY;
        for (int it = 0; it < 100; it++) {
            double mid = lo + (hi - lo) / 2;
            double areaBelow = 0;
            for (int[] sq : squares) {
                int y = sq[1], l = sq[2];
                if (mid <= y) continue;
                double h = Math.min(mid - y, l);
                areaBelow += h * l;
            }
            if (areaBelow < half) {
                lo = mid;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
}

File: Dsa-day180.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        total = sum(l * l for _, _, l in squares)
        half = total / 2
        max_y = max(y + l for _, y, l in squares)

        def area_below(h: float) -> float:
            res = 0.0
            for _, y, l in squares:
                if h <= y:
                    continue
                height = min(h - y, l)
                res += height * l
            return res

        lo, hi = 0.0, float(max_y)
        for _ in range(100):
            mid = (lo + hi) / 2
            if area_below(mid) < half:
                lo = mid
            else:
                hi = mid
        return lo